pgrx/src/spi/tuple.rs

```diff
@@ -0,0 +1,529 @@
+            Ok(PgOid::from(oid))
+        }
+    }
+
+    /// Returns column name of the 1-based `ordinal` position
+    ///
+    /// # Errors
+    ///
+    /// Returns [`Error::SpiError(SpiError::NoAttribute)`] if the specified ordinal value is out of bounds
+    /// If we have no backing tuple table a [`Error::NoTupleTable`] is returned
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if the column name at the specified ordinal position is not also
+    /// a valid UTF8 string.
+    pub fn column_name(&self, ordinal: usize) -> SpiResult<String> {
+        self.check_ordinal_bounds(ordinal)?;
+        let (_, tupdesc) = self.get_spi_tuptable()?;
+        unsafe {
+            // SAFETY:  we just got a valid tupdesc and we know ordinal is in bounds
+            let name = pg_sys::SPI_fname(tupdesc, ordinal as i32);
+
+            // SAFETY:  SPI_fname will have given us a properly allocated char* since we know
+            // the specified ordinal is in bounds
+            let str =
+                CStr::from_ptr(name).to_str().expect("column name is not value UTF8").to_string();
+
+            // SAFETY: we just asked Postgres to allocate name for us
+            pg_sys::pfree(name as *mut _);
+            Ok(str)
+        }
+    }
+
+    /// Returns the ordinal (1-based position) of the specified column name
+    ///
+    /// # Errors
+    ///
+    /// Returns [`Error::SpiError(SpiError::NoAttribute)`] if the specified column name isn't found
+    /// If we have no backing tuple table a [`Error::NoTupleTable`] is returned
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if somehow the specified name contains a null byte.
+    pub fn column_ordinal<S: AsRef<str>>(&self, name: S) -> SpiResult<usize> {
+        let (_, tupdesc) = self.get_spi_tuptable()?;
+        unsafe {
+            let name_cstr = CString::new(name.as_ref()).expect("name contained a null byte");
+            let fnumber = pg_sys::SPI_fnumber(tupdesc, name_cstr.as_ptr());
+
+            if fnumber == pg_sys::SPI_ERROR_NOATTRIBUTE {
+                Err(Error::SpiError(SpiErrorCodes::NoAttribute))
+            } else {
+                Ok(fnumber as usize)
+            }
+        }
+    }
+}
+
+impl<'conn> Iterator for SpiTupleTable<'conn> {
+    type Item = SpiHeapTupleData<'conn>;
+
+    /// # Panics
+    ///
+    /// This method will panic if for some reason the underlying heap tuple cannot be retrieved
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        self.current += 1;
+        if self.current >= self.size as isize {
+            None
+        } else {
+            assert!(self.current >= 0);
+            self.get_heap_tuple().report()
+        }
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (0, Some(self.size))
+    }
+
+    #[inline]
+    fn count(self) -> usize
+    where
+        Self: Sized,
+    {
+        self.size
+    }
+}
```

Comment:
I am not confident that it's appropriate for this type to implement Iterator itself.

Reply:
L306 seems wrong, at least. 

Why do you not think it's appropriate?  It's fairly natural to work with the results of a query as a forward-only stream of rows.

Reply:
Because it allows rewinding, so it's not very "forward-only".

Reply:
> Because it allows rewinding, so it's not very "forward-only".

It only allows rewinding because it can.  I agree that's definitely not an inherent property of "a forward-only stream of rows".

Reply:
It's usually better to implement IntoIterator if it normally allows going backwards like that.